### DISCUSSION TITLE
How do I save a gt docx table from within Quarto?

### DISCUSSION THREAD

**Discussion by robertprince** [quarto-dev/quarto-cli#11352] [Q&A] [answered]
### Description

I wish to save a gt docx table from Quarto and am receiving the following error message:

Pandoc can convert to DOCX, but not to DOC.
Error:
! pandoc document conversion failed with error 22

My r code is:

---
title: "Trial3"
format: html
editor: visual
---

{r}
# library(car)
# library(tidyverse)
# library(gtsummary)
# library(flextable)
# library(officer)
library(gt)

tab_1 <-
  gtcars |>
  dplyr::select(model, year, hp, trq) |>
  dplyr::slice(1:5) |>
  gt(rowname_col = "model") |>
  tab_stubhead(label = "car")

tab_1

tab_1 |> gtsave("tab_1.docx")

I am using Quarto version 1.6.32 on a Windows 10 pc

**Comment by mcanouil** [comment]
Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.
See <https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us>.

Could you also ensure to describe all the steps required to reproduce the error?
Remove everything not useful/required, i.e., do we need all the R packages? I believe `gt` is enough.

**Reply by Steinthal** [reply]
You're absolutely right about missing formatting and your assumption that ```gt```is enough. 
@robertprince seems to be new to Quarto (and Pandoc?) and asked the same question in the Pandoc forum. I took the code from there for my answer below.

**Reply by robertprince** [reply]
@mcanouil and @Steinthal you are both correct that I am clearly new to Quarto and Pandoc and to seeking help on this forum. My apologies if I inadvertently make mistakes as I navigate trying to resolve the problem that I am experiencing. Thank you for responding to my query.

BTW The full message that I am getting is:

Registered S3 method overwritten by 'rmarkdown':
  method         from
  print.paged_df     
Unknown output format doc
Pandoc can convert to DOCX, but not to DOC.
Error: pandoc document conversion failed with error 22

**Reply by Steinthal** [reply]
Please read the [FQA for Markdown Users](https://quarto.org/docs/faq/rmarkdown.html). My take from the FAQ is you don’t need ```rmarkdown``` when using Quarto. 

Different R packages can conflict with each other and that's what happened with your definitions. 
Hence, try to use only those packages necessary for your task (and for an example for this forum).

**Comment by Steinthal** [comment]
@robertprince, as **mcanouil** mentioned, you don't need ```flextable```and ```officer```.

This example does create the ```docx```file:

````
---
title: "Trial3"
---

```{r}
#| warning: false
#| message: false
#| echo: false
#| output: asis
library(gt)
library(car)
library(tidyverse)
library(gtsummary)

tab_1 <-
gtcars |>
dplyr::select(model, year, hp, trq) |>
dplyr::slice(1:5) |>
gt(rowname_col = "model") |>
tab_stubhead(label = "car")

tab_1

tab_1 |> gtsave("tab_1.docx")
```
````

I assume you want to use ```gtsummary```to further style the table. For the simple table you create for the test it is not necessary.

**Reply by robertprince** [reply]
@Steinthal I have tried the code on a different computer and it works. But it must be something with this particular computer that is causing the error.

**Reply by mcanouil** [reply]
Update your R packages.
Update Quarto CLI.

Also please provide the output of `quarto check`.

I can't reproduce any error.
Note that there is no reason to use `output: asis` here.

Making example as small as possible when asking for help, help others help you.
Additionally, doing this is the basics of debugging which might allow you to figure out the issue by yourself without the need to ask for external help.

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
title: "Trial3"
format: html
---

```{r}
library(gt)

tab_1 <- gtcars |>
  head() |>
  gt(rowname_col = "model") |>
  tab_stubhead(label = "car")

tab_1

tab_1 |> gtsave("tab_1.docx")
```
````

</td><td>

<img width="616" alt="image" src="https://github.com/user-attachments/assets/b35fa838-0afb-4815-8c85-c739a993bdda">

</td></tr>
</table>

**Reply by robertprince** [reply]
The output from quarto check is:

Quarto 1.6.32
[>] Checking environment information...
      Quarto cache location: C:\Users\01362160\AppData\Local\quarto
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.4.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.46.3: OK
      Typst version 0.11.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.6.32
      Path: C:\Program Files\Quarto\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Tex:  (not detected)

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....(None)

      Unable to locate an installed version of Python 3.
      Install Python 3 from https://www.python.org/downloads/

[>] Checking R installation...........OK
      Version: 4.4.2
      Path: C:/PROGRA~1/R/R-44~1.2
      LibPaths:
        - C:/Users/01362160/Documents/R/win-library/4.0
        - C:/Program Files/R/R-4.4.2/library
      knitr: 1.49
      rmarkdown: 2.29

[>] Checking Knitr engine render......OK

**Reply by mcanouil** [reply]
Thanks, so your version of Quarto CLI and your setup seems right.
Check your R packages now.

**Reply by robertprince** [reply]
My R packages are up to date.

**Reply by mcanouil** [reply]
Have you tried the following small document using `quarto render index.qmd` assuming you name the document `index.qmd`?

````qmd
---
title: "Trial3"
format: html
---

```{r}
library(gt)

tab_1 <- gtcars |>
  head() |>
  gt(rowname_col = "model") |>
  tab_stubhead(label = "car")

tab_1

tab_1 |> gtsave("tab_1.docx")
```
````

**Reply by robertprince** [reply]
@mcanouil I have rendered the file and I get the following error message:

Pandoc can convert to DOCX, but not to DOC.
Error:
! pandoc document conversion failed with error 22
Backtrace:
  1. global .main()
  2. execute(...)
  3. rmarkdown::render(...)
  4. knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet)
  5. knitr:::process_file(text, output)
     ...
 16. base::withRestarts(...)
 17. base (local) withRestartList(expr, restarts)
 18. base (local) withOneRestart(withRestartList(expr, restarts[-nr]), restarts[[nr]])
 19. base (local) docall(restart$handler, restartArgs)
 21. evaluate (local) fun(base::quote(`<smplErrr>`))

**Reply by mcanouil** [reply]
Please use `quarto render index.qmd` and provide the full output.
We can't do anything with partial outputs.

Also, please use the edition tools to format your messages.
GitHub uses markdown.
For instance, use code block for code.


For example, this is the output for the document I shared:

```
- The project is out-of-sync -- use `renv::status()` for details.
ℹ R version 4.4.2 (2024-10-31)
! Config '~/.Rprofile' was loaded!


processing file: index.qmd
output file: index.knit.md

pandoc 
  to: html
  output-file: index.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: Trial3
  
Output created: index.html
```

**Reply by robertprince** [reply]
@mcanouil It worked from the command line! I received both the html and docx files. However, when I try to run the same command from the RStudio Console I receive the following Error message:

Error: unexpected symbol in "quarto render"

The output from the command line is:

processing file: index.qmd
1/2
2/2 [unnamed-chunk-1]
output file: index.knit.md

pandoc
  to: html
  output-file: index.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png

metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: Trial3

Output created: index.html

**Reply by mcanouil** [reply]
Are you using the latest RStudio version?

In the future, please provide all necessary information as I initially requested: https://github.com/quarto-dev/quarto-cli/discussions/11352#discussioncomment-11195426

Providing a complete depiction of how to reproduce an issue with a small and I insist on small, can save a lot of time to everyone.

**Reply by robertprince** [reply]
Yes, I am.

**Reply by mcanouil** [reply]
How do you render from RStudio?
Are you sure you did not set some settings for knitr/rmarkdown?

> Error: unexpected symbol in "quarto render"

Where does this appear?

**Reply by robertprince** [reply]
I typed quarto render index.html in the RStudio Console and received the message back in the Console.

I am not sure how to go about checking the settings for knitr/rmarkdown.

**Reply by mcanouil** [reply]
To summarise, you have tried in RStudio console and in PowerShell?
Is `quarto check` output the same?

**Reply by robertprince** [reply]
@mcanouil you are correct that I have tried "quatro render index.qmd" in RStudio Console (and Terminal) and in PowerShell and the RStudio seems to give a problem while the PowerShell renders it correctly.

The quarto check is the same for both ways.

**Reply by mcanouil** [reply]
Unfortunately, I am no RStudio expert. If the check output is the same, it means there is some environment variable/settings that is causing this.
@cderv might have some insights as he is more familiar with the IDE.

Meanwhile, you could look on the RStudio GitHub repository if there is a similar issue already reported. https://github.com/rstudio/rstudio/issues

PS: you don't need to mention/ping me in every reply.

**Reply by robertprince** [reply]
BTW The file without the tab_1 |> gtsave("tab_1.docx") code at the end renders fine  on this pc. It must have something to do about saving the docx file or the gtsave command.

**Reply by mcanouil** [reply]
Maybe (as a workaround) put it in a separate code cell with `#| include: false`.

**Reply by robertprince** [reply]
@mcanouil I suspect that since I am using a business laptop I may have permissions to use tab_1 |> gtsave("tab_1.rtf") to save a rtf file but I do not to have permissions to use tab_1 |> gtsave("tab_1.docx") to save a docx file. This is extremely strange behavior. I have relocated the Quarto project folder and this allows me to save the docx file. Not ideal and I will have to engage the IT department about this behavior. Thank you so much for all your assistance.

**Reply by Steinthal** [reply]
I doubt that there are different permissions for docx and rtf files. 

If you encounter the error again, please check if the ```rmarkdown```package has been loaded. If yes, please untick the box in front of the entry.
![Screenshot 2024-11-10 at 20 19 20@2x](https://github.com/user-attachments/assets/812bbf4e-2f83-42f8-9250-4a95487b3521)

**Reply by cderv** [reply]
@robertprince about the issues you get 

```
Pandoc can convert to DOCX, but not to DOC.
Error:
! pandoc document conversion failed with error 22
```

I believe this comes from `gtsave()` which uses Pandoc internally to render the HTML table to a DOCX table. Something the wrong extension is detected. I can't reproduce, by I am pretty sure it comes from the `rmarkdown::pandoc_convert()` calls done by **gt** 

> when I try to run the same command from the RStudio Console I receive the following Error message:
> Error: unexpected symbol in "quarto render"

`quarto` is a Command Line Interface (CLI) tool. It means it is to be called in a Terminal. It is not a R function or package. So `quarto render` in the R console means nothing. You need to use it in the Terminal from RStudio IDE. Or use the **quarto** R package throuhg `quarto::quarto_render()` that will call the CLI for your in the background. 

> I suspect that since I am using a business laptop I may have permissions to use tab_1 |> gtsave("tab_1.rtf") to save a rtf file but I do not to have permissions to use tab_1 |> gtsave("tab_1.docx") to save a docx file.

It is quite odd, and I don't have a clue on why this PC is not allowing this. Possible something the in the project, or the environment, that makes **gt** consider the wrong file extension. 🤷

**Comment by Luccan97** [comment, marked as answer]
I was having the same error and I found the answer here: 
"The problem is that gtsave have problem with overwriting .docx documents (that's not the case with .rtf)" 

so... change the file name or delete the old one.

https://github.com/rstudio/gt/issues/1726
